pkg/sdk/tasks_impl_gen.go

```diff
@@ -66,40 +68,38 @@ func (v *tasks) Execute(ctx context.Context, request *ExecuteTaskRequest) error
 // GetRootTasks is a way to get all root tasks for the given tasks.
 // Snowflake does not have (yet) a method to do it without traversing the task graph manually.
 // Task DAG should have a single root but this is checked when the root task is being resumed; that's why we return here multiple roots.
-// Cycles should not be possible in a task DAG but it is checked when the root task is being resumed; that's why this method has to be cycle-proof.
-// TODO [SNOW-884987]: handle cycles
+// Cycles should not be possible in a task DAG, but it is checked when the root task is being resumed; that's why this method has to be cycle-proof.
 func GetRootTasks(v Tasks, ctx context.Context, id SchemaObjectIdentifier) ([]Task, error) {
-	task, err := v.ShowByID(ctx, id)
-	if err != nil {
-		return nil, err
-	}
+	tasksToExamine := collections.NewQueue[SchemaObjectIdentifier]()
+	alreadyExaminedTasksNames := make([]string, 0)
+	rootTasks := make([]Task, 0)
 
-	predecessors := task.Predecessors
-	// no predecessors mean this is a root task
-	if len(predecessors) == 0 {
-		return []Task{*task}, nil
-	}
+	tasksToExamine.Push(id)
+
+	for tasksToExamine.Head() != nil {
+		current := tasksToExamine.Pop()
 
-	rootTasks := make([]Task, 0, len(predecessors))
-	for _, predecessor := range predecessors {
-		predecessorTasks, err := GetRootTasks(v, ctx, predecessor)
+		if slices.Contains(alreadyExaminedTasksNames, current.Name()) {
+			continue
+		}
+
+		task, err := v.ShowByID(ctx, *current)
 		if err != nil {
-			return nil, fmt.Errorf("unable to get predecessors for task %s err = %w", predecessor.FullyQualifiedName(), err)
+			return nil, err
 		}
-		rootTasks = append(rootTasks, predecessorTasks...)
-	}
 
-	// TODO [SNOW-884987]: extract unique function in our collection helper (if cycle-proof algorithm still needs it)
-	keys := make(map[string]bool)
-	uniqueRootTasks := make([]Task, 0, len(rootTasks))
-	for _, rootTask := range rootTasks {
-		if _, exists := keys[rootTask.ID().FullyQualifiedName()]; !exists {
-			keys[rootTask.ID().FullyQualifiedName()] = true
-			uniqueRootTasks = append(uniqueRootTasks, rootTask)
+		predecessors := task.Predecessors
+		if len(predecessors) == 0 {
+			rootTasks = append(rootTasks, *task)
+		} else {
+			for _, p := range predecessors {
+				tasksToExamine.Push(p)
+			}
 		}
+		alreadyExaminedTasksNames = append(alreadyExaminedTasksNames, current.Name())
```

Comment:
I think using current.Name() is an issue (Before this FullyQualifiedName was being returned in the array of task names)? After upgrading to the current release we're getting failures when creating or applying changes to DAG's of tasks related to the root task being unable to be found.

Reply:
Hey @rmargarint-nydig. The `alreadyExaminedTasksNames` is built using only `Name()`s so I don't think that this is the issue.

If you are experiencing any problems with tasks, please create an issue with the description and the smallest possible sample showing the problems you are experiencing. Thanks!